realtek: add DGS-1210-28 factory image

luizluca opened the pull request:
DGS-1210 switches supports dual image, which each image composed of a
kernel and a rootfs partition. For image1, kernel and rootfs are in
sequence. The current OpenWrt image (written using serial console), uses
those partitions together as the firmware partition, ignoring the
partition division. The current OEM u-boot fails to validate image1 but
it will only trigger firmware recovery if both image1 and image2 fails,
and it does not switch the boot image in case one of them fails the
check.

The OEM factory image is composed of contatenated blocks of data, each
one prefixed with a 0x40-byte header, with each block representing a
partition (kernel, rootfs). dlink-imghdr.py is a tool to create that
header and it is a clean-room reimplementation of imghdr present in the
DGS-1210-28-GPL package. Except for the header, OEM firmware does not
validate the image content.

The OpenWrt factory image mimics the OEM image by cutting the
kernel+rootfs firmware at the exactly size of the OEM kernel partition
and packing it as "the kernel partition" and the rest of the kernel and
the rootfs as "the rootfs partition". It will only work if written to
image1 because image2 has a sysinfo partition between kernel2 and
rootfs2, cutting the kernel code in the middle.

Steps to install:

1) switch to image2 (containing an OEM image), using web or these CLI
   commands:
   - config firmware image_id 2 boot_up
   - reboot
2) flash the factory_only_for_image1.bin to image1. OEM web (v6.30.016)
   is crashing for any upload (ssh keys, firmware), even applying OEM
   firmwares. This CLI commands can upload a new firmware to the other
   image location (not used to boot):
   - download firmware_fromTFTP <tftpserver> factory_only_for_image1.bin
   - config firmware image_id 1 boot_up
   - reboot

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://openwrt.org/submitting-patches

Please remove this message before posting the pull request.

luizluca committed:
realtek: add DGS-1210-28 factory image

DGS-1210 switches support dual image, with each image composed of a
kernel and a rootfs partition. For image1, kernel and rootfs are in
sequence. The current OpenWrt image (written using a serial console),
uses those partitions together as the firmware partition, ignoring the
partition division. The current OEM u-boot fails to validate image1 but
it will only trigger firmware recovery if both image1 and image2 fail,
and it does not switch the boot image in case one of them fails the
check.

The OEM factory image is composed of concatenated blocks of data, each
one prefixed with a 0x40-byte cameo header. A normal OEM firmware will
have two of these blocks (kernel, rootfs). The OEM firmware only checks
the header before writing unconditionally the data (except the header)
to the correspoding partition.

The OpenWrt factory image mimics the OEM image by cutting the
kernel+rootfs firmware at the exact size of the OEM kernel partition
and packing it as "the kernel partition" and the rest of the kernel and
the rootfs as "the rootfs partition". It will only work if written to
image1 because image2 has a sysinfo partition between kernel2 and
rootfs2, cutting the kernel code in the middle.

Steps to install:

1) switch to image2 (containing an OEM image), using web or these CLI
   commands:
   - config firmware image_id 2 boot_up
   - reboot
2) flash the factory_image1.bin to image1. OEM web (v6.30.016)
   is crashing for any upload (ssh keys, firmware), even applying OEM
   firmwares. These CLI commands can upload a new firmware to the other
   image location (not used to boot):
   - download firmware_fromTFTP <tftpserver> factory_image1.bin
   - config firmware image_id 1 boot_up
   - reboot

To debrick the device, you'll need serial access. If you want to
recover to an OpenWrt, you can replay the serial installation
instructions. For returning to the original firmware, press ESC during
the boot to trigger the emergency firmware recovery procedure. After
that, use D-Link Network Assistant v2.0.2.4 to flash a new firmware.

The device documentation does describe that holding RESET for 12s
trigger the firmware recovery. However, the latest shipped U-Boot
"2011.12.(2.1.5.67086)-Candidate1" from "Aug 24 2021 - 17:33:09" cannot
trigger that from a cold boot. In fact, any U-Boot procedure that relies
on the RESET button, like reset settings, will only work if started from
a running original firmware. That, in practice, cancels the benefit of
having two images and a firmware recovery procedure (if you are not
consider dual-booting OpenWrt).

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

musashino205 commented:
The header doesn't seem to be D-Link specific. I think it is used in devices manufactured by Cameo, including APRESIA ApresiaLightGS series.

my note: https://memo205.wordpress.com/2021/09/20/dgs-1210-28-f1-%e3%83%95%e3%82%a1%e3%83%bc%e3%83%a0%e3%82%a6%e3%82%a7%e3%82%a2%e3%83%a1%e3%83%a2/

svanheule commented on target/linux/realtek/image/rtl838x.mk:
```diff
@@ -39,6 +42,11 @@ TARGET_DEVICES += d-link_dgs-1210-20
+  FACTORY_IMG_BOARD_VERSION := 32
+  IMAGES += factory_only_for_image1.bin
+  IMAGE/factory_only_for_image1.bin := append-kernel | pad-to 64k | append-rootfs | pad-rootfs | \
+	          pad-to 16 | check-size | dlink-header
```
`dlink-header` will make the file bigger, no? Unless the vendor FW will unpack the blobs and the padded file is to one actually written to flash (not clear to me from the commit message), `dlink-header` must be called before `check-size`.

luizluca replied:
It drops the headers before flashing. IMAGE_SIZE is exactly the sum of the kernel and the rootfs size. I'll try to explain it better in the commit message.

luizluca replied:
Added some extra info to commit message

svanheule commented on scripts/dlink-imghdr.py:
```diff
@@ -0,0 +1,69 @@
+parser.add_argument('signature')
+parser.add_argument('partition', type=int, choices=range(0,10), metavar="partition=[0-9]",help="partition id")
+parser.add_argument('costumer_signature', type=int, choices=range(0,10), metavar="costumer_signature=[0-9]",help="costumer signature")
+parser.add_argument('board_version', type=int, choices=range(0,2**32), metavar="board_version=[0-4294967295]",help="costumer signature")
```
These lines are too long, please wrap them.

svanheule commented on target/linux/realtek/image/Makefile:
```diff
@@ -16,6 +16,21 @@ define Build/zyxel-vers
+		${FACTORY_IMG_CUSTOMER_SIGNATURE} \
+		${FACTORY_IMG_BOARD_VERSION} \
+		$(KERNEL_LOADADDR); \
+        $(SCRIPT_DIR)/dlink-imghdr.py $@.rootfs_part $@.rootfs_part.hex \
```
Mixing spaces and tabs

luizluca replied:
Done

svanheule commented on target/linux/realtek/image/Makefile:
```diff
@@ -16,6 +16,21 @@ define Build/zyxel-vers
 
+define Build/dlink-header
+	head --bytes $(FACTORY_KERNEL_PART_SIZE) $@ > $@.kernel_part; \
+	tail --bytes +$(shell echo $$(($(FACTORY_KERNEL_PART_SIZE)+1))) $@ > $@.rootfs_part; \
```
Would this work too? (untested)
```Makefile
dd bs=$(1) count=1 $(2) > $(2).kernel_part;
dd bs=$(1) skip=1 $(2) > $(2).rootfs_part;
...
```

luizluca replied:
I switched to dd

svanheule reviewed (commented):
Please split this into two commits to:
* add the image generator,
* enable the -factory images for the D-Link devices.

luizluca commented on scripts/dlink-imghdr.py:
```diff
@@ -0,0 +1,69 @@
+
+MODEL_LEN=20
+SIGNATURE_LEN=16
+LINUXLOAD_LEN=10
```
OK. Not a big fan on 80 col limit but I'm used to it.

svanheule replied:
As far as I'm concerned 100 characters is also fine. But ~135 chars might be a bit too much :)

svanheule replied:
Could you also use [spaces](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#spaces) for assignments and between function arguments? You currently have some assignments without spaces around the operator, and some with.

luizluca replied:
OK. Done

luizluca commented:
> The header doesn't seem to be D-Link specific. I think it is used in devices manufactured by Cameo, including APRESIA ApresiaLightGS series.
> 
> my note: https://memo205.wordpress.com/2021/09/20/dgs-1210-28-f1-%e3%83%95%e3%82%a1%e3%83%bc%e3%83%a0%e3%82%a6%e3%82%a7%e3%82%a2%e3%83%a1%e3%83%a2/

I wish I had that info before ;-) My google search didn't match it probably it is written in Japanese.

luizluca commented on target/linux/realtek/image/rtl838x.mk:
```diff
@@ -14,6 +14,9 @@ TARGET_DEVICES += allnet_all-sg8208m
   IMAGE_SIZE := 13824k
+  FACTORY_KERNEL_PART := 2
+  FACTORY_KERNEL_PART_SIZE := 1572864
+  FACTORY_ROOTFS_PART := 3
```
I'll use CAMEO, as it is not really a dlink thing.

svanheule replied:
I didn't mean you needed to move them from `Device/d-link_dgs-1210` to the device specific build recipe. The parameters that these devices have in common can stay here.

You need to add the `DLINK_` and `CAMEO_` variable names to `DECIVE_VARS` in `target/linux/realtek/image/Makefile`:
```Makefile
DEVICE_VARS += ZYXEL_VERS DLINK_KERNEL_PART_SIZE ...
```

luizluca committed:
scripts: add cameo image header generator

The cameo header is a 0x40-byte header used by D-Link DGS 1210 switches
and Apresia ApresiaLightGS series. cameo-imghdr.py is a clean-room
reimplementation of imghdr present in the DGS-1210-28-GPL package.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

svanheule commented on target/linux/realtek/image/rtl838x.mk:
```diff
@@ -39,6 +39,14 @@ TARGET_DEVICES += d-link_dgs-1210-20
+  CAMEO_ROOTFS_PART := 3
+  CAMEO_CUSTOMER_SIGNATURE := 2
+  CAMEO_BOARD_VERSION := 32
+  IMAGES += factory_only_for_image1.bin
```
Deviations from just the `factory.bin` name are normally used to support (device or firmware) revisions that require different factory images, not to prevent users from shooting themselves in the foot. Some users can (and probably will) still just ignore the file name and flash to image2.

Could you also document how to debrick the device in the commit message, in case an image is flashed to image2? It appears this device was added without a proper description, so it would be good to have this info.

luizluca replied:
> Deviations from just the `factory.bin` name are normally used to support (device or firmware) revisions that require different factory images, not to prevent users from shooting themselves in the foot. Some users can (and probably will) still just ignore the file name and flash to image2.

I know but we always try hard to avoid it.
 
> Could you also document how to debrick the device in the commit message, in case an image is flashed to image2? It appears this device was added without a proper description, so it would be good to have this info.

It's complicated. The device does have an emergency recovery from u-boot, but we can only trigger it from a working original image. It looks like the reset button does not work from a cold reboot. It is not just for this recovery: not even config reset works from a cold reboot (or from an OpenWrt boot). The only recovery possible is by triggering it from a serial console by pressing ESC. From there, there easiest way to recovery is using the D-Link Network Assistant v2.0.2.4. If you want to do it manually, you'll need to cut the firmware where the kernel blob+cameo tag ends, remove header from both parts and flash each one to the corresponding partition. I'll add these info to the commit message (and wiki).

> I'm not entirely convinced about the `factory_only_for_image1.bin` name, but maybe `factory_image1.bin` would be an option.

I'm fine with factory_image1.bin, if it will be accepted.

luizluca commented:
I hope I've dealt with all comments.

svanheule commented on target/linux/realtek/image/Makefile:
```diff
@@ -16,6 +17,21 @@ define Build/zyxel-vers
+        dd if=$@ bs=$(DLINK_KERNEL_PART_SIZE) count=1 of=$@.kernel_part; \
+        dd if=$@ bs=$(DLINK_KERNEL_PART_SIZE) skip=1 of=$@.rootfs_part; \
+        $(SCRIPT_DIR)/cameo-imghdr.py $@.kernel_part $@.kernel_part.hex \
+                ${DEVICE_MODEL} os ${CAMEO_KERNEL_PART} \
```
Just one more detail, and then we're done.

For variable substitution, we normally use parentheses instead of braces. Both are valid, and you've used both here, but I think we should be consistent. You should also add quotes for `"$(DEVICE_MODEL)"`, since that value may contain spaces.

luizluca replied:
Done!

luizluca commented:
> The commit message for "[realtek: add DGS-1210-28 factory image](https://github.com/openwrt/openwrt/pull/10113/commits/0825b33a7833aa9ba99c4838931d746040b87874)" still uses the "factory_only_for_image1" suffix. This should be updated to match the actual image suffix.

Thanks for your reviews! It's done. Once this gets in, I'll submit the DGS-1210-52 PR.

svanheule commented:
There was still a copy-paste error in the help text for the `board_version` argument, but I fixed that and merged the patches into master. Thanks!
